sklearn/ensemble/iforest.py

```diff
@@ -250,17 +250,23 @@ def decision_function(self, X):
         """
         # code structure from ForestClassifier/predict_proba
         # Check data
-        X = self.estimators_[0]._validate_X_predict(X, check_input=True)
+        X = check_array(X, accept_sparse='csr')
         n_samples = X.shape[0]
 
         n_samples_leaf = np.zeros((n_samples, self.n_estimators), order="f")
         depths = np.zeros((n_samples, self.n_estimators), order="f")
 
-        for i, tree in enumerate(self.estimators_):
-            leaves_index = tree.apply(X)
-            node_indicator = tree.decision_path(X)
+        for i, (tree, features) in enumerate(zip(self.estimators_,
+                                                 self.estimators_features_)):
+            if len(features) == X.shape[0]:
```

Comment:
And it would be better to use `self._max_features`. You can test it outside the loop:
```
if self._max_features == X.shape[1]:
        random_projection = True
```
Then in the loop, use `random_projection` instead of `len(features) == X.shape[1]`